src/rpi_macaddr.c

```diff
@@ -14,88 +14,96 @@
-                // read parts are little endian, swap for easier printf
-                uint64_t macaddr_le = ((uint64_t)(mac1 & 0xFFFF) << 32) | mac0;
-                uint64_t macaddr = bswap_64(macaddr_le) >> 16;
-
-                // apply magic numbers
-                macaddr = macaddr ^ (xor_magic & 0xFFFFFF);
-                // make sure the additive magic number does not affect the OUI part of the MAC
-                macaddr = (macaddr & 0xFFFFFF000000) | ((macaddr + add_magic) & 0xFFFFFF);
-
-                snprintf(dst, MAX_SERIALNUMBER_LEN, "%06llx", macaddr);
-                success = true;
-            }
-        }
+    uint32_t words[VC_MBOX_GET_MAC_ADDRESS_BUFFER_SIZE / 4] =
+        {VC_MBOX_GET_MAC_ADDRESS_BUFFER_SIZE, 0, VC_MBOX_GET_MAC_ADDRESS, VC_MBOX_GET_MAC_ADDRESS_VALUE_SIZE, 0, 0, 0, 0};
+    int rc = ioctl(fd, VC_MBOX_IOCTL_PROPERTY, &words);
+    if (rc >= 0 &&
+        words[0] == VC_MBOX_GET_MAC_ADDRESS_BUFFER_SIZE &&
+        words[1] == 0x80000000 &&
+        words[2] == VC_MBOX_GET_MAC_ADDRESS &&
+        words[3] == VC_MBOX_GET_MAC_ADDRESS_VALUE_SIZE) {
+        memcpy(mac_address, &words[5], 6);
+        success = true;
     }
 
-    // popen() malloc'd lineptr buffer, we have to free it after we're done with it
-    if (lineptr)
-        free(lineptr);
-
-    pclose(vc_file);
-
+    close(fd);
     return success;
 }
 
-
-bool rpi_wlan0_macaddr_id(const struct boardid_options *options, char *buffer)
+static bool rpi_macaddr_id(char *buffer, bool wlan0)
 {
-    return rpi_macaddr_vc(buffer, RPI_WLAN_XOR_MAGIC, 0);
-}
+    uint8_t mac_address[6];
+    if (!vc_read_mac_address(mac_address))
+        return false;
+
+    if (wlan0) {
+        // The wlan0 MAC address is computed differently depending on the
+        // version of the RPi.  In particular, if the VC reports the OUI
+        // B8:27:EB, the NIC part of the MAC address comes from the serial
+        // number. The WiFi MAC is programmed with the NIC part XORed with
+        // 0x555555.
+        //
+        // The Raspberry Pi 4 (and presumably newer ones) do not derive their
+        // MAC addresses from the serial number and instead have them
+        // programmed to OTP memory. They do not have an OUI of B8:27:EB. The
+        // currently known ones have an OUI of DC:A6:32, but the RPi Foundation
+        // also owns E4:5F:01. If the MAC isn't programmed, the user can
+        // program it. For these devices, the procedure is to add 1 to the NIC
+        // part.
+        //
+        // Disclaimer: I couldn't find anything official to say that any of
+        // this is true, but it seems to be so far.
+
+        if (mac_address[3] == 0xb8 && mac_address[4] == 0x27 && mac_address[5] == 0xeb) {
+            mac_address[3] ^= 0xff;
+            mac_address[4] ^= 0xff;
+            mac_address[5] ^= 0xff;
+        } else {
+            int v = (mac_address[3] << 16) | (mac_address[3] << 8) | mac_address[5];
```

Comment:
Fixing the middle assignment as well as changing temp var to `unsigned` because strictly speaking overflow isn't well-defined for signed integers.
```suggestion
            unsigned int v = (mac_address[3] << 16) | (mac_address[4] << 8) | mac_address[5];
```